### DISCUSSION TITLE
Avalonia .axaml

### DISCUSSION THREAD

**Discussion by Fxplorer** [ionide/ionide-vscode-fsharp#2002] [General]
I guess I am a bit obscure in that I am trying to use the [Avalonia UI](https://avaloniaui.net) framework in conjunction with F# in vscode on linux ([LMDE6](https://linuxmint.com/download_lmde.php)). My problem arises in ionide when I open a project and then go to the solution view (needed for ordering). The Avalonia .axaml files are not listed. I am not sure if WPF .xaml files show there or not. 
I have the Avalonia extension installed, which will give the axaml previewer when in file mode. But with the ordering requirement, i tend to do F# in the solution view. 

I am no where advanced in F# enough to provide help in implementing this function yet. Is there any insight into how this might be updated in ionide to show these .axaml files? I am wondering how hard this would be to do. the files are listed in the .fsproj file. I would assume that ionide is just ignoring all but known file types.

**Comment by baronfel** [comment]
Do you have a sample project or repo that illustrates the behavior? The solution view is driven by data that we pull from MSBuild, so it's _probably_ a matter of extending the set of MSBuild Items that we look for in FsAutoComplete - a repro would help us verify this.

**Reply by Fxplorer** [reply]
repo https://github.com/Fxplorer/AvaloniaIonide

**Reply by Fxplorer** [reply]
here is a video showing this also

[https://www.dropbox.com/scl/fi/c1ueeea3a8gypq430qde8/2024-04-16-13-08-47.mp4?rlkey=43dfizt70wyxqos1s14zkjag0&dl=0](https://www.dropbox.com/scl/fi/c1ueeea3a8gypq430qde8/2024-04-16-13-08-47.mp4?rlkey=43dfizt70wyxqos1s14zkjag0&dl=0)

**Reply by baronfel** [reply]
So this is because these are a kind of MSBuild Item called `AvaloniaXaml`. The Avalonia build targets define these, but they are not `Compile` items. There's also no linkage that I can find (in terms of MSBuild metadata) between the AvaloniaXaml items and their matching axaml.fs items - usually there's _something_ that we can build one.

We could potentially have a set of additional items that we return in addition to Compile items (AvaloniaXaml, etc), but I would probably want some additional work on our side to detect+nest the files together.

**Reply by baronfel** [reply]
As soon as I said that I found the thing I was looking for:

* The Avalonia targets [do set `DependentUpon`](https://github.com/AvaloniaUI/Avalonia/blob/1fe718a4ec21599257d48b86dfb98a5c03ba70bd/packages/Avalonia/AvaloniaBuildTasks.targets#L21-L32), which is the signal that we'd need to create this association
* BUT! They only do this for `Compile` items
[* BUT! They never add axaml/paml files to the `Compile` items, only AndroidXaml items](https://github.com/AvaloniaUI/Avalonia/blob/1fe718a4ec21599257d48b86dfb98a5c03ba70bd/packages/Avalonia/AvaloniaBuildTasks.props#L7-L10) (probably because they would blow up the C#/F# compilers)

So, I'd suggest two things:
* An issue should get logged on Avalonia to figure out what they want to do with DependentUpon - I think they should look at adding this metadata to the AvaloniaXaml MSBuild Items
* We could potentially automatically collect and forward along any Item that is `DependentUpon` to a Compile item

cc @TheAngryByrd / @Krzysztof-Cieslak for thoughts.
